### PR TITLE
[BE]: pyupgrade Python to 3.8 - remove extraneous parentheses only

### PR DESCRIPTION
**Pull request by bryant1410**
This change removes all redundant/extraneous parentheses. It was done by running:

```bash
ruff check --target-version=py38 --select=UP034 --fix --verbose benchmarks/ caffe/ test/ tools/ torch/
```

I reviewed them all and they make sense to me.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @desertfire @anijain2305

### CODE REVIEW

**Comment by bryant1410** [comment]
(BTW, not sure what "[BE]" is in the title, but I tried to imitate similar PRs)

File: test/test_torch.py [hunk 1]
```diff
@@ -6614,9 +6614,9 @@ def test_parsing_intlist(self):
 
         # fail parse with > 1 element variables
         self.assertRaises(TypeError, lambda: torch.ones(torch.tensor(3, 3)))
-        self.assertRaises(TypeError, lambda: torch.ones((torch.tensor(3, 3))))
+        self.assertRaises(TypeError, lambda: torch.ones(torch.tensor(3, 3)))
+        self.assertRaises(TypeError, lambda: torch.ones(np.array(3, 3)))
         self.assertRaises(TypeError, lambda: torch.ones(np.array(3, 3)))
-        self.assertRaises(TypeError, lambda: torch.ones((np.array(3, 3))))
```
Comment (janeyx99) [inline test/test_torch.py:6619]: This one seems intentionally a tuple, so it would be good to ensure this is not breaking intentionality.

Also, if this is a safe change, lines 6618 and 6619 look the same now, so one can be removed.

**Comment by janeyx99** [comment]
I've given the changes a glance and they seem overall good. My concern is that in some of the test files, I do not know if the tests intend to pass tuples vs flattened values and if we are accidentally losing coverage this way. Would appreciate another reviewer's thoughts.

File: test/test_torch.py (hunk 1 above)
Comment (bryant1410) [reply test/test_torch.py:6619]: `(np.array(3, 3))` is not a tuple. It needs a trailing comma to be a tuple (`(np.array(3, 3),)`). So maybe it's a bug from the test?

Comment (bryant1410) [reply test/test_torch.py:6619]: So maybe we should fix these bugs? After having the confirmation that they are actual bugs. Because it'd be weird to have this duplicate line otherwise

**Comment by bryant1410** [comment]
> Okay, in that case, my "taking a step back" question would be: is ruff the new standard for lint we want to enforce and is it enforced in lintrunner today? If so, we should ensure lint will prevent future instances of breaking this specific rule. cc'ing @huydhn for review about our lint standards.

@XuehaiPan maybe you also know about this?

**Comment by bryant1410** [comment]
I'm thinking that many of these parentheses are for readability reasons (and that maybe I should undo?) and that maybe we don't wanna enforce a general rule but maybe there are some subrules that could be enforced?

File: test/test_torch.py (hunk 1 above)
Comment (bryant1410) [reply test/test_torch.py:6619]: Okay, my interpretation is that these intended to be tuples, so I fixed them now.

**Comment by bryant1410** [comment]
Alright, so I did the following changes:
* Added a rule to automatically check at lint time the fix of this PR (I guess it'll run on CI?).
* Fixed more files that I found after I ran the rule.
* On second thoughts, I don't think there are files that need the parentheses detected as a way to improve code clarity.

It should be ready for review again.

**Comment by kit1980** [comment]
Not sure if we want to touch caffe2.
It's not really maintained and we don't have tests in CI, so we can miss errors.
Plus updating it may give a false impression it's maintained.

**Comment by kit1980** [comment]
> I've given the changes a glance and they seem overall good. My concern is that in some of the test files, I do not know if the tests intend to pass tuples vs flattened values and if we are accidentally losing coverage this way. Would appreciate another reviewer's thoughts.

One note is that in Python `(c)` is not a tuple containing `c`, it's just `c`.
Comma makes a tuple, so `(c, )` is a tuple.
I don't think any of the changes change a tuple to non-tuple (and that's would be a bug in ruff).

File: pyproject.toml [hunk 2]
```diff
@@ -63,6 +63,7 @@ select = [
     "E",
     "F",
     "SIM1",
+    "UP034",
```
Comment (XuehaiPan) [inline pyproject.toml:66]: I think it would be better to select `UP0` and then add unfixed rules to `ignore`. Then we can gradually remove the ignored rules one by one rather than add them to `select` list. Just we do for `SIM1`.

Comment (bryant1410) [reply pyproject.toml:66]: What if new `UP0*` rules appear with new ruff versions, and all of a sudden CI starts failing?

**Comment by bryant1410** [comment]
> Not sure if we want to touch caffe2.
> It's not really maintained and we don't have tests in CI, so we can miss errors.
> Plus updating it may give a false impression it's maintained.

Should I undo those?

File: pyproject.toml (hunk 2 above)
Comment (bryant1410) [reply pyproject.toml:66]: Oh, okay!

**Comment by huydhn** [comment]
We add ruff linter recently, so using it to format code is fine IMO.  But let's do it via lintrunner https://github.com/pytorch/pytorch/blob/main/.lintrunner.toml#L999-L1023.  There are some directories excluded there, which would address @kit1980 concern about touching `caffe2` as that directory is excluded

**Comment by bryant1410** [comment]
> We add ruff linter recently, so using it to format code is fine IMO. But let's do it via lintrunner https://github.com/pytorch/pytorch/blob/main/.lintrunner.toml#L999-L1023. There are some directories excluded there, which would address @kit1980 concern about touching `caffe2` as that directory is excluded

Do I need to make changes? My understanding is that the lines you pointed out load `pyproject.toml`, which is the file I changed.

**Comment by huydhn** [comment]
> Do I need to make changes? My understanding is that the lines you pointed out load `pyproject.toml`, which is the file I changed.

Ah, my point is that you have already updated `pyproject.toml`, you can just run `linrunner -a` to apply ruff according to the rule defined in https://github.com/pytorch/pytorch/blob/main/.lintrunner.toml#L999-L1023, no need to manually run the ruff command you mention in the PR description (nothing is wrong with that command, it's just that some directories should be excluded like caffe2)

**Review by malfet — APPROVED** [review]
LGTM, but please consider splitting it into smaller PRs per folder: one for `torch`, one for `test`, and final one that enforces it in `pyproject.toml`, as otherwise it would be very susceptible to land races.

File: .github/scripts/trymerge.py [hunk 3]
```diff
@@ -1261,9 +1261,9 @@ def find_matching_merge_rule(
             if num_matching_files > reject_reason_score:
                 reject_reason_score = num_matching_files
                 reject_reason = "\n".join(
-                    f"Not all files match rule `{rule_name}`."
-                    f"{num_matching_files} files matched, but there are still non-matching files:"
-                    f"{','.join(non_matching_files[:5])}{', ...' if len(non_matching_files) > 5 else ''}"
+                    f"Not all files match rule `{rule_name}`.",
+                    f"{num_matching_files} files matched, but there are still non-matching files:",
+                    f"{','.join(non_matching_files[:5])}{', ...' if len(non_matching_files) > 5 else ''}",
```
Comment (malfet) [inline .github/scripts/trymerge.py]: If `ruff` does it, it's wrong here, as `join` does not take multiple arguments, but only one (at least on 3.9):
```
% python3.9 -c "print('-'.join('a', 'b', 'c'))"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: str.join() takes exactly one argument (3 given)
```

Comment (bryant1410) [reply .github/scripts/trymerge.py]: No, it was me failing to solve a merge conflict with a commit from a PR that originated from a conversation from this PR: https://github.com/pytorch/pytorch/pull/101606#discussion_r1197146340. Note you made a review on an outdated version (see the following commit that fixes it).

Comment (bryant1410) [reply .github/scripts/trymerge.py]: Oh, it was a conversation w you actually. But it wasn't ruff, it was my incorrect merge conflict resolution.

File: pyproject.toml [hunk 4]
```diff
@@ -63,6 +98,7 @@ select = [
...
```
Comment (XuehaiPan) [inline pyproject.toml:101]:
```suggestion
    "UP0",
```

Comment (bryant1410) [reply pyproject.toml:101]: Oops, I though I had changed that. But why not just "UP"?

Comment (XuehaiPan) [reply pyproject.toml:101]: > But why not just "UP"?

I think it could be `UP`. I.e., select all `UP` rules.

**Comment by bryant1410** [comment]
> LGTM, but please consider splitting it into smaller PRs per folder: one for `torch`, one for `test`, and final one that enforces it in `pyproject.toml`, as otherwise it would be very susceptible to land races.

@malfet should I create a PR per folder then? That sounds like a lot since there are 10 folders touched here.

**Comment by bryant1410** [comment]
> > LGTM, but please consider splitting it into smaller PRs per folder: one for `torch`, one for `test`, and final one that enforces it in `pyproject.toml`, as otherwise it would be very susceptible to land races.
> 
> @malfet should I create a PR per folder then? That sounds like a lot since there are 10 folders touched here.

Friendly ping!

**Comment by malfet** [comment]
@bryant1410 If you can create several smaller PR I'll be more than happy to review/land them.

**Comment by kit1980** [comment]
@bryant1410 I wonder if you want to return to this and submit smaller PRs?

**Comment by bryant1410** [comment]
Haven't had time but will try to find some time soon

**Comment by bryant1410** [comment]
Okay, IIUC, the following PRs should be created:

* [ ] .github/
* [ ] benchmarks/
* [ ] caffe2/
* [ ] docs/
* [ ] functorch/
* [ ] pyproject.toml
* [ ] test/
* [ ] third_party/
* [ ] tools/
* [ ] torch/

**Review by Skylion007 — APPROVED** [review]

**Comment by justinchuby** [comment]
All ruff `UP` rules are enabled by this stack https://github.com/pytorch/pytorch/pull/105437. I think we are covered.

**Comment by bryant1410** [comment]
> All ruff `UP` rules are enabled by this stack #105437. I think we are covered.

**Comment by bryant1410** [comment]
I'm gonna reopen it to see the diff with the current overlapping changes, as they're merged.

**Comment by justinchuby** [comment]
The stack is all merged. Please feel free to rebase and move this PR forward

File: pyproject.toml [hunk 5]
```diff
@@ -55,6 +55,41 @@ ignore = [
...
+    "UP005",
+    "UP006",
+    "UP007",
+    "UP008",
+    "UP009",
+    "UP010",
+    "UP011",
+    "UP012",
+    "UP013",
+    "UP014",
```
Comment (justinchuby) [inline pyproject.toml:70]: Be sure to undo the changes to pyproject and take it from main

File: torch/utils/jit/__init__.py [hunk 6]
```diff
@@ -1 +0,0 @@
-
```
Comment (justinchuby) [inline torch/utils/jit/__init__.py:1]: ?

Comment (bryant1410) [reply torch/utils/jit/__init__.py:1]: It just removes an empty line in that file

Comment (justinchuby) [reply torch/utils/jit/__init__.py:1]: Should the file have at least one empty line? (end with `\n`)

Comment (justinchuby) [reply torch/utils/jit/__init__.py:1]: Although as long as lintrunner is happy we should be good

**Comment by bryant1410** [comment]
Okay, I merged with the latest (I guess in this context is the same as rebasing since it can be rebased down the road).

The remaining changes I see in this PR are:

* Applying the rule to the directory `caffe/`
* Applying the rule to the directory `third_party/`
* It applies 2 missing usages of the rule (why are they undetected?) to files under `test/`.
* It fixes a test from `test/test_torch.py`, which seems to have been wrong for long (maybe always?). Because it seems that it was intended to assert the usage of both with and without a tuple.
* It removes an unnecessary "\n" char from an otherwise empty `__init__.py` file

What should we do with these changes?

File: torch/utils/jit/__init__.py (hunk 6 above)
Comment (bryant1410) [reply torch/utils/jit/__init__.py:1]: It's not common in Python, and also not in this codebase. See the output of:

```bash
find torch/ -name __init__.py -exec ls -l {} \;
```

It's the only "empty" one like that.

File: test/jit/test_module_containers.py [hunk 7]
```diff
@@ -643,7 +643,7 @@ def forward(self, x):
 
         n._reconstruct(m._c)
 
-        inp = torch.rand((3))
+        inp = torch.rand(3)
```
Comment (justinchuby) [inline test/jit/test_module_containers.py:646]: This path was excluded. That's why it was undetected. But this change lgtm

File: test/inductor/test_torchinductor_dynamic_shapes.py [hunk 8]
```diff
@@ -60,7 +60,7 @@
         ("cuda"), is_skip=True
     )
     test_failures["test_batch_norm_2d_dynamic_shapes"] = TestFailure(
-        ("cuda"), is_skip=True
+      "cuda", is_skip=True
```
Comment (justinchuby) [inline test/inductor/test_torchinductor_dynamic_shapes.py:63]: Not sure why the stack missed this

File: test/jit/test_module_containers.py (hunk 7 above)
Comment (bryant1410) [reply test/jit/test_module_containers.py:646]: I'm curious, why is it excluded?

File: test/inductor/test_torchinductor_dynamic_shapes.py (hunk 8 above)
Comment (bryant1410) [reply test/inductor/test_torchinductor_dynamic_shapes.py:63]: I see the previous lines weren't changed. Maybe I had done this manually for some reason; I can't remember.

I'm gonna also change the previous lines.

File: test/jit/test_module_containers.py (hunk 7 above)
Comment (justinchuby) [reply test/jit/test_module_containers.py:646]: Actually maybe we should keep? This is because the jit tests test the jit compiler's ability to understand python code, and it needs to support valid python in these test cases. It is excluded in `.lintrunner` with comments

File: test/inductor/test_torchinductor_dynamic_shapes.py [hunk 9]
```diff
@@ -54,13 +54,13 @@
         ("cpu", "cuda"), is_skip=True
     )
     test_failures["test_convolution3_dynamic_shapes"] = TestFailure(
-        ("cuda"), is_skip=True
+        "cuda", is_skip=True
```
[the remark this answers is not in the document]
Comment (bryant1410) [reply test/inductor/test_torchinductor_dynamic_shapes.py:57]: But it seems to work fine without being a tuple?

File: test/jit/test_module_containers.py (hunk 7 above)
Comment (bryant1410) [reply test/jit/test_module_containers.py:646]: So, what should I do?

Comment (justinchuby) [reply test/jit/test_module_containers.py:646]: I will let maybe @albanD or @malfet to comment on this

Comment (albanD) [reply test/jit/test_module_containers.py:646]: This particular line is not testing jit behavior. You can keep this change here.

File: test/inductor/test_torchinductor_dynamic_shapes.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (albanD) [reply test/inductor/test_torchinductor_dynamic_shapes.py:57]: It is used as `if suffix in tf.suffixes` so I guess we're lucky and making it a string or a list of string has a similar behavior.

Let's not change any semantic in this cleanup PR though and if we think it needs solving, either open a follow up PR or an issue.

File: test/test_torch.py [hunk 10]
```diff
@@ -6704,9 +6704,9 @@ def test_parsing_intlist(self):
 
         # fail parse with > 1 element variables
         self.assertRaises(TypeError, lambda: torch.ones(torch.tensor(3, 3)))
-        self.assertRaises(TypeError, lambda: torch.ones(torch.tensor(3, 3)))
-        self.assertRaises(TypeError, lambda: torch.ones(np.array(3, 3)))
+        self.assertRaises(TypeError, lambda: torch.ones((torch.tensor(3, 3),)))
```
Comment (albanD) [inline test/test_torch.py:6707]: Wait this is semantically not the same thing! What suggested this change?

**Review by albanD — COMMENTED** [review]
Sounds mostly ok except the test_torch change!

File: test/test_torch.py (hunk 10 above)
Comment (bryant1410) [reply test/test_torch.py:6707]: I think we had talked about this before in this PR. I did it manually.

But it makes sense, right? Because, otherwise, there are duplicate asserts.

File: test/inductor/test_torchinductor_dynamic_shapes.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (bryant1410) [reply test/inductor/test_torchinductor_dynamic_shapes.py:57]: I think that to comply with the API, it should be a tuple, yeah. So maybe we should open up a follow up PR.

**Comment by bryant1410** [comment]
Can somebody help me understand why did the failing test fail?

File: test/test_torch.py (hunk 10 above)
Comment (janeyx99) [reply test/test_torch.py:6707]: @bryant1410 I think you likely meant to deduplicate lines 6706 and 6707 NOT 6707 and 6708

**Comment by janeyx99** [comment]
The current failing test looks to be an existing unstable failure--you can tell by looking at the pytorchbot comment: 
image

File: test/test_torch.py (hunk 10 above)
Comment (bryant1410) [reply test/test_torch.py:6707]: The diff is confusing but the end result is fine IMHO.

Before this change, there were 2 pairs of duplicate lines. They, in turn, before a recent [BE] PR merge, had been originally syntactically different but semantically the same because they had redundant parentheses which the mentioned PR removed.

So, my change is not to remove the duplicates but to fix what I interpret the original intention was, which is to assert that the function call works both with and without a single-element tuple (for each pair of duplicates).

File: test/inductor/test_torchinductor_dynamic_shapes.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kit1980) [reply test/inductor/test_torchinductor_dynamic_shapes.py:57]: @bryant1410 
`("cuda")` is not a tuple, it's a string "cuda".
`("cuda", )` is a tuple.

[the remark this answers is not in the document]
Comment (bryant1410) [reply test/inductor/test_torchinductor_dynamic_shapes.py:57]: Yes, that's clear to me. I'm saying that the expected type is a tuple and in consequence this should be changed (to a tuple).

File: test/test_torch.py (hunk 10 above)
Comment (albanD) [reply test/test_torch.py:6707]: Hey!
I would have two things to say:
- We should NOT do semantic changes in pyupgrade/lint PRs. If we want to do this, this should be a separate PR.
- Note that 6700 above basically tests this case already. So I don't think we need to add this.

Given this, in this PR, we should just remove these lines.
